### PR TITLE
Add a new config object for compactConfig

### LINKED ISSUE #5242: New Config Object for compactConfig
```CoordinatorDynamicConfig``` is designed to accept simple configuration values rather than complicated JSON objects. #5102 introduces a new configuration, ```compactConfig```, which is implemented as a part of ```CoordinatorDynamicConfig``` in that PR. Since ```compactConfig``` is a complicated JSON object, we need another new config object for it rather than using ```CoordinatorDynamicConfig```. 

Also, ```CoordinatorDynamicConfig``` should be documented as it is designed to accept simple values.

### PR DESCRIPTION
**Pull request by jihoonson**
Fixes #5242.

---
This change is [Reviewable](https://reviewable.io/reviews/druid-io/druid/5264)

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java [hunk 1]
```diff
@@ -21,158 +21,127 @@
...
-    return targetCompactionSizeBytes;
+    return new CoordinatorCompactionConfig(ImmutableList.of(), null, null);
   }
 
-  @JsonProperty
-  public int getNumTargetCompactionSegments()
+  @JsonCreator
+  public CoordinatorCompactionConfig(
+      @JsonProperty("compactionConfigs") List<DataSourceCompactionConfig> compactionConfigs,
+      @JsonProperty("compactionTaskSlotRatio") Double compactionTaskSlotRatio,
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java]: Annotate `@Nullable`

File: server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java [hunk 2]
```diff
@@ -21,158 +21,127 @@
...
+      @JsonProperty("maxCompactionTaskSlots") Integer maxCompactionTaskSlots
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java]: And this

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 3]
```diff
@@ -331,8 +294,6 @@ public static Builder builder()
     private static final boolean DEFAULT_EMIT_BALANCING_STATS = false;
     private static final boolean DEFAULT_KILL_ALL_DATA_SOURCES = false;
     private static final int DEFAULT_MAX_SEGMENTS_IN_NODE_LOADING_QUEUE = 0;
-    private static final double DEFAULT_COMPACTION_TASK_RATIO = 0.1;
-    private static final int DEFAILT_MAX_COMPACTION_TASK_SLOTS = Integer.MAX_VALUE;
 
     private Long millisToWaitBeforeDeleting;
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Please annotate properties and parameters in this class `@Nullable` as appropriate.

File: server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+import io.druid.client.indexing.ClientCompactQueryTuningConfig;
+import org.joda.time.Period;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Objects;
+
+public class DataSourceCompactionConfig
+{
+  // should be synchronized with Tasks.DEFAULT_MERGE_TASK_PRIORITY
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java]: Please make this a javadoc comment with a link

File: server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  private static final int DEFAULT_COMPACTION_TASK_PRIORITY = 25;
+  private static final long DEFAULT_TARGET_COMPACTION_SIZE_BYTES = 800 * 1024 * 1024; // 800MB
+  private static final int DEFAULT_NUM_TARGET_COMPACTION_SEGMENTS = 150;
+  private static final Period DEFAULT_SKIP_OFFSET_FROM_LATEST = new Period("P1D");
+
+  private final String dataSource;
+  private final int taskPriority;
+  private final long targetCompactionSizeBytes;
+  // The number of compaction segments is limited because the byte size of a serialized task spec is limited by
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java]: And here

File: server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  // RemoteTaskRunnerConfig.maxZnodeBytes.
+  private final int numTargetCompactionSegments;
+  private final Period skipOffsetFromLatest;
+  private final ClientCompactQueryTuningConfig tuningConfig;
+  private final Map<String, Object> taskContext;
+
+  @JsonCreator
+  public DataSourceCompactionConfig(
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("taskPriority") Integer taskPriority,
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java]: `@Nullable`, and 3 params after this

File: server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    } else {
+      newCompactionConfig = new CoordinatorCompactionConfig(null, compactionTaskSlotRatio, maxCompactionTaskSlots);
+    }
+
+    if (!manager.set(
+        CoordinatorCompactionConfig.CONFIG_KEY,
+        newCompactionConfig,
+        new AuditInfo(author, comment, req.getRemoteAddr())
+    )) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java]: minor nit: is it possible to add some message about the failure cause here ? Maybe something on lines - Failed to set config...

File: server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    } else {
+      newCompactionConfig = CoordinatorCompactionConfig.from(ImmutableList.of(newConfig));
+    }
+
+    if (!manager.set(
+        CoordinatorCompactionConfig.CONFIG_KEY,
+        newCompactionConfig,
+        new AuditInfo(author, comment, req.getRemoteAddr())
+    )) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java]: minor nit: same, please add more details for error if possible.

**Comment by gianm** [comment]
Marking as 0.13.0 since #5242 is marked as such; and this patch is responsive to that issue.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java (hunk 1 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java]: Added.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java (hunk 2 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java]: Added.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 3 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Added.

File: server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java (hunk 4 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java]: `Tasks` is in a different module and can't be linked.

File: server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java (hunk 5 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java]: Same. Can't be linked.

File: server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java (hunk 6 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java]: Added.

File: server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java (hunk 7 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java]: Added.

File: server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java (hunk 8 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java]: Added.

**Comment by jihoonson** [comment]
@leventov @nishantmonu51 @gianm thanks for the review. Addressed your comments.

@gianm there are already some docs. Here are the links.

- https://github.com/druid-io/druid/blob/master/docs/content/design/coordinator.md#compacting-segments
- https://github.com/druid-io/druid/blob/master/docs/content/design/coordinator.md#http-endpoints

Do you think something more should be added?

**Comment by gianm** [comment]
@jihoonson, doesn't this patch add a new endpoint: `/druid/coordinator/v1/config/compaction` and remove the compaction config from the dynamic config (where it used to be)? I meant reflecting those changes in the docs.

**Comment by jihoonson** [comment]
@gianm actually the API endpoint is not changed. I only added a new config object for compaction config. Do you think the API endpoint should be changed as well?

**Comment by gianm** [comment]
Sorry, I was being really dumb. I didn't realize the endpoint already existed through a different code path.

**Comment by jihoonson** [comment]
@gianm oh, sorry. I finally got your point. Fixed docs according to the changed APIs.

**Comment by jihoonson** [comment]
@gianm thanks. Fixed.

**Pull request merged**